File: przyklady-zlozonosci/cpp/node.cpp
#include <iostream>
using namespace std;

#include "node.hpp"

// constructors
Node::Node() {
  index = -1;
  value = 0;
  next = nullptr;
  showIndex = true;
}

Node::Node(Node &node) {
  index = node.index;
  value = node.value;
  next = node.next;
  showIndex = true;
}

Node::Node(int providedValue) {
  index = -1;
  value = providedValue;
  next = nullptr;
  showIndex = true;
}

Node::Node(int providedValue, Node &providedNext) {
  index = -1;
  value = providedValue;
  next = &providedNext;
  showIndex = true;
}

Node::Node(int providedValue, int providedIndex) {
  index = providedIndex;
  value = providedValue;
  next = nullptr;
  showIndex = true;
}

Node::Node(int providedIndex, int providedValue, Node &providedNext) {
  index = providedIndex;
  value = providedValue;
  next = &providedNext;
  showIndex = true;
}

// destructors
Node::~Node() { delete next; }

// setters and getters
void Node::setValue(int newValue) { value = newValue; }

int Node::getValue() { return value; }

void Node::setIndex(int newIndex) { index = newIndex; }

int Node::getIndex() { return index; }

Node *Node::getNextPtr() { return next; }

void Node::resetNextPtr() { next = nullptr; }

void Node::setNextPtr(Node &newNext) { next = &newNext; }

void Node::setShowIndex(bool newShowIndex) { showIndex = newShowIndex; }

// other funcs
void Node::print() {
  if (index != -1 && showIndex) {
    cout << "[" << index << "] ";
  }

  cout << value << endl;
}

File: complexity-examples/python-bad/node.py
from __future__ import annotations


class Node:
    def __init__(self, value: int = 0, index: int = -1, next: Node = None, showIndex: bool = False) -> Node:
        self._value = value
        self._index = index
        self._next = next
        self._showIndex = showIndex

    def __str__(self) -> str:
        return self._print()

    def __repr__(self) -> str:
        return self._print()

    # private funcs
    def _print(self) -> str:
        return_str = ''

        if self._index != -1 and self._showIndex:
            return_str += f'[{self._index}] '

        return return_str + f'{self._value}'

    # setters & getters
    def setValue(self, newValue: int) -> None:
        self._value = newValue

    def getValue(self) -> int:
        return self._value

    def setIndex(self, newIndex: int) -> None:
        self._index = newIndex

    def getIndex(self) -> int:
        return self._index

    def setShowIndex(self, newShowIndex: bool) -> None:
        self._showIndex = newShowIndex

    def resetNext(self) -> None:
        self._next = None

    def setNext(self, newNext: Node) -> None:
        self._next = newNext

    def getNext(self) -> Node:
        return self._next

    # public funcs
    def print(self) -> str:
        return self._print()


if __name__ == "__main__":
    pass
